lib/Sema/CSSimplify.cpp

```diff
@@ -298,6 +299,47 @@ matchCallArguments(SmallVectorImpl<AnyFunctionType::Param> &args,
     if (numClaimedArgs == numArgs)
       return None;
 
+    /// When we determine which argument is bound to unlabeled parameter,
+    /// consider still unbounded parameter which is prior to current parameter.
+    /// In order not to intersect binding position that remaining parameter will
+    /// bind later, skip a unlabeled argument as much as it can.
+    ///
+    /// For example:
+    /// @code
+    /// func f(aa: Int, _ bb: Int) {}
+    /// f(0, 1)
+    /// @endcode
+    /// Choice argument[1] for parameter[1] so that parameter[0] will be bounded
+    /// to argument[0] later.
+    ///
+    /// Because variadics parameter can be bounded with more than one arguments,
+    /// they don't this.
+    if (paramLabel.empty() && paramIdx && !forVariadic &&
+        !params[*paramIdx].isVariadic()) {
+      unsigned unboundedParamCount = 0;
+      for (unsigned pi = 0; pi < *paramIdx; pi++) {
+        if (parameterBindings[pi].empty()) {
+          if (params[pi].isVariadic() || paramInfo.hasDefaultArgument(pi))
+            continue;
+
+          unboundedParamCount++;
+        }
+      }
+
+      unsigned keepedArgCount = 0;
+      for (unsigned ai = nextArgIdx; ai < numArgs; ai++) {
+        if (claimedArgs[ai])
+          continue;
+
+        nextArgIdx = std::max(nextArgIdx, ai);
+
+        if (keepedArgCount >= unboundedParamCount) {
```

Comment:
Why this is relationship between unclaimed arguments and parameters important that way?

Reply:
This is exactly important.

`unboundedParamCount` represents how many arguments should be skipped.
A number of skipping is derived from unclaimed parameters.
`keepedArgCount` is just counter to process skipping.

I explain with two examples.

## example1

```swift
func f(aa: Int, _ bb: Int) {}

f(0, 1)
```

I use following definitions for explanation:

- `bindNextParameter.1` is `bindNextParameter` called from line 604.
- `bindNextParameter.2` is `bindNextParameter` called from line 689.

With example1, current behavior is following.

```
execution:
  bindNextParameter.1(paramIdx=0, nextArgIdx=0):
    params[aa:] => fail

  bindNextParameter.1(paramIdx=1, nextArgIdx=0):
    params[_ bb:] => `0` -- (t1)

  bindNextParameter.2(paramIdx=0, nextArgIdx=0):
    params[aa:] => `1`

bindings:
  aa: => `1`
  _ bb: => `0` (crossing!)

errors:
  - missing label for `aa:`
```

 I want to shift binding target for `_ bb:` from `0` to `1` at (t1).
 So it needs to skip *one* argument which is `0`.
 The number *one* is derived from `unboundedParamCount`.
 `keepedArgCount` is counter which represents number of skipped arguments during skipping loop.

With example1, patched behavior is following:

```
execution:
  bindNextParameter.1(paramIdx=0, nextArgIdx=0):
    params[aa:] => fail

  bindNextParameter.1(paramIdx=1, nextArgIdx=0):
    adjustment:
      unboundedParamCount = 1
      keepedArgCount = 1
      nextArgIdx = 1
    params[_ bb:] => `1`

  bindNextParameter.2(paramIdx=0, nextArgIdx=0):
    params[aa:] => `0`

bindings:
  aa: => `0`
  _ bb: => `1`  

errors:
  - missing label for `aa:`
```

Labeling errors are same.
But it avoids crossed bindings.

## example2

```swift
func f(aa: Int, bb: Int, _ cc: Int, dd: Int = 0) {}

f(0, 1, 2)
```

With example1, current behavior is following.

```
execution:
  bindNextParameter.1(paramIdx=0, nextArgIdx=0):
    params[aa:] => fail

  bindNextParameter.1(paramIdx=1, nextArgIdx=0):
    params[bb:] => fail

  bindNextParameter.1(paramIdx=2, nextArgIdx=0):
    params[_ cc:] => `0` -- (t2)

  bindNextParameter.1(paramIdx=3, nextArgIdx=0):
    params[dd:] => fail

  bindNextParameter.2(paramIdx=0, nextArgIdx=0):
    params[aa:] => `1`

  bindNextParameter.2(paramIdx=1, nextArgIdx=0):
    params[bb:] => `2`

bindings:
  aa: => `1`
  bb: => `2`
  _ cc: => `0` (crossing!)
  dd: => none

errors:
  - missing argument label for `aa:`
  - missing argument label for `bb:`
```

I want to shift binding target for `_ cc:` from `0` to `2` at (t2).
So it needs to skip *two* arguments which are `0` and `1`.
The number *two* is derived from `unboundedParamCount`.
 `keepedArgCount` is counter which represents number of skipped arguments during skipping loop.

With example1, patched behavior is following:

```
execution:
  bindNextParameter.1(paramIdx=0, nextArgIdx=0):
    params[aa:] => fail

  bindNextParameter.1(paramIdx=1, nextArgIdx=0):
    params[bb:] => fail

  bindNextParameter.1(paramIdx=2, nextArgIdx=0):
    adjustment:
      unboundedParamCount = 2
      keepedArgCount = 2
      nextArgIdx = 2
    params[_ cc:] => `2`

  bindNextParameter.2(paramIdx=0, nextArgIdx=0):
    params[aa:] => `0`

  bindNextParameter.2(paramIdx=1, nextArgIdx=1):
    params[bb:] => `1`

bindings:
  aa: => `0`
  bb: => `1`
  _ cc: => `2`
  dd: => none

errors:
  - missing argument label for `aa:`
  - missing argument label for `bb:`
```

In this case, labeling errors are same.
But it avoids crossed bindings.

Reply:
I understand the intent here but the problem is that `unboundedParamCount` is logically unrelated to `keepedArgCount` because unclaimed parameters could be scattered all over the call, their quantity shouldn't affect argument consideration... It's convenient that it works for some examples but I'm pretty sure there are examples made worse with these changes.

Reply:
@xedin 
I can understand what you say that possibility of case happen bad effect.
I created new patch at #32082 and revise this idea.
New algorithm do similar thing in stage 2 what I call.
This is run in separated group by labeled argument binding in stage 1.
So it is well structured and I don't think bad effect more.